https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

In [3]:
'''Trains a simple convnet on the MNIST dataset.
Gets to 99.25% test accuracy after 12 epochs
(there is still a lot of margin for parameter tuning).
16 seconds per epoch on a GRID K520 GPU.
'''

from __future__ import print_function
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K

In [4]:
batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [5]:
print(K.image_data_format())
if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

channels_last


In [6]:
input_shape

(28, 28, 1)

In [7]:
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples


In [8]:
# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
print('y_train shape', y_train.shape)
print(y_train.shape[0], 'train g.t. samples')
print(y_test.shape[0], 'test g.t. samples')

y_train shape (60000, 10)
60000 train g.t. samples
10000 test g.t. samples


In [9]:
model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


In [10]:
model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])

In [11]:
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))

Instructions for updating:
Use tf.cast instead.
Train on 60000 samples, validate on 10000 samples
Epoch 1/12
60000/60000 [==============================] - 8s 126us/step - loss: 0.2594 - acc: 0.9197 - val_loss: 0.0560 - val_acc: 0.9823
Epoch 2/12
60000/60000 [==============================] - 5s 76us/step - loss: 0.0862 - acc: 0.9740 - val_loss: 0.0380 - val_acc: 0.9871
Epoch 3/12
60000/60000 [==============================] - 4s 74us/step - loss: 0.0653 - acc: 0.9808 - val_loss: 0.0378 - val_acc: 0.9870
Epoch 4/12
60000/60000 [==============================] - 4s 75us/step - loss: 0.0546 - acc: 0.9839 - val_loss: 0.0314 - val_acc: 0.9893
Epoch 5/12
60000/60000 [==============================] - 5s 78us/step - loss: 0.0459 - acc: 0.9855 - val_loss: 0.0307 - val_acc: 0.9898
Epoch 6/12
60000/60000 [==============================] - 5s 78us/step - loss: 0.0402 - acc: 0.9874 - val_loss: 0.0321 - val_acc: 0.9908
Epoch 7/12
60000/60000 [==============================] - 4s 74us/step - loss: 

In [14]:
'https://stackoverflow.com/questions/35792278/how-to-find-number-of-parameters-of-a-keras-model'
def size_model(model): # Compute number of params in a model (the actual number of floats)
    import numpy as np
    #return sum([np.prod(K.get_value(w).shape) for w in model.trainable_weights])
    W = [K.get_value(w).shape for w in model.trainable_weights]
    print(W)
    total = sum([np.prod(w) for w in W])
    return total
                 
    

In [15]:
size_model(model)

[(3, 3, 1, 32), (32,), (3, 3, 32, 64), (64,), (9216, 128), (128,), (128, 10), (10,)]


1199882

In [16]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.02847562966905498
Test accuracy: 0.991


#作業１：是否有對資料做標準化 (normalization)？ 如果有，在哪幾行？

1.1 有作標準化,
1.2 請搜索以下兩行的對應行數!
x_train /= 255
x_test /= 255


#作業２：使用的優化器 Optimizer 為何？ 

Adadelta

#作業３：模型總共疊了幾層卷積層？ 

只有第一層是捲積層

#作業４：模型的參數量是多少？
1199882
